rasterio/windows.py

```diff
@@ -303,10 +302,71 @@ def shape(window, height=-1, width=-1):
     values in the window.
     """
     (a, b), (c, d) = evaluate(window, height, width)
-    return b-a, d-c
+    return (b - a, d - c)
 
 
 def window_index(window):
     # "window_" is necessary here to redundancy to disambiguate
     # from transform.get_index and src.index
     return tuple(slice(*w) for w in window)
+
+
+class Window(tuple):
+    """Windows are rectangular subsets of rasters"""
+
+    __slots__ = ()
+    _fields = ('col_off', 'row_off', 'num_cols', 'num_rows')
+
+    def __new__(cls, col_off=0, row_off=0, num_cols=0, num_rows=0):
+        """Create new instance of Window."""
+        return tuple.__new__(
+            cls,
+            ((row_off, row_off + num_rows), (col_off, col_off + num_cols)))
+
+    def __repr__(self):
+        """Return a nicely formatted representation string"""
+        return (
+            "Window(col_off={self.col_off}, row_off={self.row_off}, "
+            "num_cols={self.num_cols}, num_rows={self.num_rows})").format(
+                self=self)
+
+    def __getnewargs__(self):
+        'Return self as a plain tuple.  Used by copy and pickle.'
+        return self.flattened()
+
+    def flattened(self):
+        """Return a flattened tuple, ordered as _fields."""
+        return (self[1][0], self[0][0],
+                self[1][1] - self[1][0], self[0][1] - self[0][0])
+
+    def asdict(self):
+        'Return a new OrderedDict which maps field names to their values.'
+        return collections.OrderedDict(zip(self._fields, self.flattened()))
+
+    @classmethod
+    def from_ranges(cls, row_range, col_range):
+        return cls(col_range[0], row_range[0],
+                   col_range[1] - col_range[0], row_range[1] - row_range[0])
+
+    @classmethod
+    def from_bounds(cls, left, bottom, right, top, transform,
+            height=None, width=None, boundless=False, precision=6):
+        return from_bounds(left, bottom, right, top, transform,
+                           height=height, width=width, boundless=boundless,
+                           precision=precision)
+
```

Comment:
> one and only one obvious way to do it

There's good precedent for breaking this convention. Specifically numpy's ndarray methods are effectively just a wrappers around top-level functions that take ndarrays as the first arg.

```
>>> import numpy as np

>>> a = np.random.random(1000)

>>> np.sum(a) == a.sum()
True
```